File: 2023/UICPC Regional/Round #15/solutions/dungeondawdler/jo.py
F='NESW'
D=[(-1,0),(0,1),(1,0),(0,-1)]
add=lambda X, Y: (X[0] + Y[0], X[1] + Y[1])
sub=lambda X, Y: (X[0] - Y[0], X[1] - Y[1])
step=lambda X, i: add(X, D[i])
rev=lambda A:map(lambda x:(x+2)%4, A[::-1])

class State:
    def __init__(S, U):
        S.x = (0,0)
        S.path = []
        S.G = {S.x:'.'}
        S.seen = set([S.x])
        S.visited = set()
        S.visit(U)

    def visit(S, U):
        S.visited.add(S.x)
        for i,v in enumerate(U):
            p = step(S.x, i)
            if not p in S.seen:
                S.seen.add(p)
                S.G[p] = v
        if S.G[S.x] == '.':
            S.G[S.x] = ":"

    def print_ans(S):
        c0=c1=r0=r1=0
        M=dict([(k,'.'if v==':'else v)for k,v in S.G.items()])
        for r,c in M.keys():
            c0,c1,r0,r1=min(c0,c),max(c1,c),min(r0,r),max(r1,r)
        print 'done\n%d %d'%(r1-r0+1,c1-c0+1)
        for r in range(r0, r1+1):
            print ''.join([M.get((r,c), '#') for c in range(c0, c1+1)])

    def go(S, s):
        print F[s]
        ln=raw_input().split()
        S.x = step(S.x, s)
        S.visit(ln[0])
        if ln[1] == 'trap':
            S.trap = S.x
            return State(ln[2])

    def search(S, no=[]):
        while True:
            s = None
            for i in range(4):
                p = step(S.x, i)
                if p not in no and p not in S.visited and S.G[p] != '#':
                    s = i
                    break
            if s is None:
                if not S.path: break
                s = (S.path[-1]+2)%4
                del S.path[-1]
            else:
                S.path += [s]
            succ = S.go(s)
            if succ: return succ

    def walk(S, path):
        for s in path:
            p = step(S.x, s)
            if S.G.get(p, '#') == '#': break
            S.path += [s]
            succ = S.go(s)
            if succ: return succ

def merge(A, B, off):
    for rc in B:
        A[add(rc, off)] = B[rc]

def solve(S, A, ab, pab, ba, pba):
    T = [ab]
    Am = A.search(no=T)
    if Am:
        T += [A.trap]
        Am.search(no=T)
        Bm = Am.walk(pab)
    else:
        Am, Bm = A, A.walk(pab)
    B = Bm.search(no=[ba])
    if B:
        B.search(no=[ba, Bm.trap])
        Bm = B
    Bm.walk(pba)
    if len(T) == 2:
        O = sub(T[1], ba)
        S0 = sub(T[1], S.trap)
    else:
        for rc,v in Bm.G.iteritems():
            if v == '.':
                O = sub(T[0], rc)
                break
        S0 = sub(add(O, ba), S.trap)
    merge(Am.G, Bm.G, O)
    Am.G[add(O, ba)] = 'A'
    Am.G[T[0]] = 'B'
    Am.G[(0,0)] = 'a'
    Am.G[O] = 'b'
    Am.G[S0] = 'S'
    Am.print_ans()

def solve_in(S, wx, wy, T, P):
    disc = wx.path
    b = wy.walk(disc)
    if b:
        a = b.search(no=[wx.trap])
        ab = a.walk(P[0])
        bx = ab.walk(disc)
        if bx:
            ax = bx.walk(b.path)
            solve(S, ax, T[0], P[0], b.trap, b.path)
        else:
            ab.walk(rev(ab.path))
            solve(S, ab, T[1], P[1], b.trap, b.path)
    else:# ? => P[0] = A->A
        wy.walk(rev(wy.path))
        solve(S, wy, T[1], P[1], wx.trap, wx.path)

def solve_out(S, wx, T, P, i):
    disc = P[i]
    wy = wx.search()
    z = wy.walk(disc)
    if not z:
        z = wy.search(no=[wx.trap])
        w1 = z.walk(P[0])
        w2 = w1.walk(disc)
        if w2:
            if disc == P[1]:
                w2 = w2.walk(wy.path)
            solve(S, w2, T[1], P[1], wy.trap, wy.path)
        else:
            z = w1.walk(rev(w1.path) + wy.path)
            solve(S, z, T[0], P[0], wy.trap, wy.path)
    else:
        w1 = z.walk(disc)
        if w1:
            solve(S, w1, T[1-i], P[1-i], wx.trap, wx.path)
        else:
            z = z.walk(rev(z.path) + wx.path)
            solve(S, z, T[i], P[i], wx.trap, wx.path)

S = State(raw_input().split()[0])
A = S.search()
if not A:
    S.G[S.x] = 'S'
    S.print_ans()
else:
    B = A.search()
    C = B.walk(A.path)
    if C and B.path == A.path:
        wx = C.search(no=[A.trap])
        if not wx:
            wx = C.walk(A.path)
            wy = wx.search(no=[A.trap])
            if not wy:
                tA = A.trap
                O = sub(tA, S.trap)
                merge(C.G, S.G, O)
                C.G[tA] = 'A'
                C.G[(0,0)] = 'a'
                C.G[O] = 'S'
                C.print_ans()
            else:
                solve_in(S, wx, wy, [A.trap, C.trap], [A.path, C.path])
        else:
            T, P = [A.trap, C.trap], [A.path, C.path]
            for i in [0,1]:
                wy = wx.walk(P[i])
                if not wy or wx.path != P[i]:
                    break
                wx = wy
            if not wy:
                solve_out(S, wx, T, P, i)
            elif wx.path != P[i]:
                solve_in(S, wx, wy, T, P)
    else:
        if not C:
            C = B.search()
        wx = C.walk(A.path)
        if wx and A.path == C.path:
            wy = wx.search(no=[B.trap])
            if not wy: wy = wx
            z = wy.walk(B.path)
            solve(S, z, A.trap, A.path, B.trap, B.path)
        else:
            if wx: C = wx
            z = C.search(no=[B.trap])
            solve(S, z, A.trap, A.path, C.trap, C.path)

File: 2023/UICPC Regional/Round #15/solutions/dungeondawdler/JO.java
import java.util.*;
import java.io.*;
import static java.lang.Math.*;
import static java.util.Collections.*;
public class JO {
  static final Vec Z = new Vec(5000,5000);
  static final char[] F = {'N','E','S','W'};
  static final Vec[] D = {new Vec(-1,0),new Vec(0,1),new Vec(1,0),new Vec(0,-1)};
  static final Set<Integer> nil = emptySet();
  public static void main(String[] args) throws Exception {
    try (BufferedReader in = new BufferedReader(new InputStreamReader(System.in))) {
      State S, A, B, C, wx, wy, z;
      S = new State(in.readLine());
      A = S.search(in, nil);
      if (A == null) {
        S.G.put(S.x.e, 'S');
        S.printAns();
      } else {
        B = A.search(in, nil);
        C = B.walk(in, A.path);
        if (C != null && B.path.equals(A.path)) {
          wx = C.search(in, singleton(A.trap.e));
          if (wx == null) {
              wx = C.walk(in, A.path);
              wy = wx.search(in, singleton(A.trap.e));
              if (wy == null) {
                  Vec tA = A.trap;
                  Vec O = tA.sub(S.trap);
                  C.merge(S, O);
                  C.G.put(tA.e, 'A');
                  C.G.put(Z.e, 'a');
                  C.G.put(Z.add(O).e, 'S');
                  C.printAns();
              } else {
                List<List<Integer>> P = new ArrayList<>();
                P.add(A.path);
                P.add(C.path);
                solveIn(in, S, wx, wy, Arrays.asList(new Vec[] {A.trap, C.trap}), P);
              }
          } else {
            List<Vec> T = Arrays.asList(new Vec[] {A.trap, C.trap});
            List<List<Integer>> P = new ArrayList<>();
            P.add(A.path);
            P.add(C.path);
            int i = 0;
            wy = wx.walk(in, P.get(i));
            if (wy != null && wx.path.equals(P.get(i))) {
              wx = wy;
              wy = wx.walk(in, P.get(++i));
              if (wy != null && wx.path.equals(P.get(i)))
                wx = wy;
            }
            if (wy == null)
                solveOut(in, S, wx, T, P, i);
            else if (!wx.path.equals(P.get(i)))
                solveIn(in, S, wx, wy, T, P);
          }
        } else {
          if (C == null) C = B.search(in, nil);
          wx = C.walk(in, A.path);
          if (wx != null && A.path.equals(C.path)) {
            wy = wx.search(in, singleton(B.trap.e));
            if (wy == null) wy = wx;
            z = wy.walk(in, B.path);
            solve(in, S, z, A.trap, A.path, B.trap, B.path);
          } else {
            if (wx != null) C = wx;
            z = C.search(in, singleton(B.trap.e));
            solve(in, S, z, A.trap, A.path, C.trap, C.path);
          }
        }
      }
    }
  }

  static void solve(BufferedReader in, State S, State A, Vec ab, List<Integer> pab, Vec ba, List<Integer> pba) throws Exception {
    List<Integer> T = new ArrayList<>(),
      Tb = new ArrayList<>();
    T.add(ab.e);
    State Am, B, Bm;
    Am = A.search(in, T);
    if (Am != null) {
      T.add(A.trap.e);
      Am.search(in, T);
      Bm = Am.walk(in, pab);
    } else {
      Am = A;
      Bm = A.walk(in, pab);
    }
    Tb.add(ba.e);
    B = Bm.search(in, Tb);
    if (B != null) {
      Tb.add(Bm.trap.e);
      B.search(in, Tb);
      Bm = B;
    }
    Bm.walk(in, pba);
    Vec O = null, S0;
    if (T.size() == 2) {
      S0 = new Vec(T.get(1)).sub(S.trap);
      O = new Vec(T.get(1)).sub(ba);
    } else {
      for (int k : Bm.G.keySet()) {
        if (Bm.G.get(k) == '.') {
          O = new Vec(T.get(0)).sub(new Vec(k));
          break;
        }
      }
      S0 = O.add(ba).sub(S.trap);
    }
    Am.merge(Bm, O);
    Am.G.put(O.add(ba).e, 'A');
    Am.G.put(T.get(0), 'B');
    Am.G.put(Z.e, 'a');
    Am.G.put(Z.add(O).e, 'b');
    Am.G.put(Z.add(S0).e, 'S');
    Am.printAns();
  }

  static void solveIn(BufferedReader in, State S, State wx, State wy, List<Vec> T,
      List<List<Integer>> P) throws Exception {
    State a, b, ab, ax, bx;
    List<Integer> disc = wx.path;
    b = wy.walk(in, disc);
    if (b != null) {
      a = b.search(in, singleton(wx.trap.e));
      ab = a.walk(in, P.get(0));
      bx = ab.walk(in, disc);
      if (bx != null) {
        ax = bx.walk(in, b.path);
        solve(in, S, ax, T.get(0), P.get(0), b.trap, b.path);
      } else {
        ab.walk(in, rev(ab.path));
        solve(in, S, ab, T.get(1), P.get(1), b.trap, b.path);
      }
    } else {
      wy.walk(in, rev(wy.path));
      solve(in, S, wy, T.get(1), P.get(1), wx.trap, wx.path);
    }
  }

  static void solveOut(BufferedReader in, State S, State wx, List<Vec> T,
      List<List<Integer>> P, int i) throws Exception {
    State a, b, w1, w2, z, wy;
    List<Integer> disc = P.get(i);
    wy = wx.search(in, nil);
    z = wy.walk(in, disc);
    if (z == null) {
      z = wy.search(in, singleton(wx.trap.e));
      w1 = z.walk(in, P.get(0));
      w2 = w1.walk(in, disc);
      if (w2 != null) {
        if (disc.equals(P.get(1)))
          w2 = w2.walk(in, wy.path);
        solve(in, S, w2, T.get(1), P.get(1), wy.trap, wy.path);
      } else {
        w1.walk(in, rev(w1.path));
        z = w1.walk(in, wy.path);
        solve(in, S, z, T.get(0), P.get(0), wy.trap, wy.path);
      }
    } else {
      w1 = z.walk(in, disc);
      if (w1 != null) {
        solve(in, S, w1, T.get(1-i), P.get(1-i), wx.trap, wx.path);
      } else {
        z.walk(in, rev(z.path));
        z = z.walk(in, wx.path);
        solve(in, S, z, T.get(i), P.get(i), wx.trap, wx.path);
      }
    }
  }

  static List<Integer> rev(List<Integer> P) {
    List<Integer> rev = new ArrayList<>();
    for (int i = P.size()-1; i >= 0; --i) {
      rev.add((P.get(i)+2)%4);
    }
    return rev;
  }
}

class Vec {
  static final int R = (1<<14)-1;
  final int r,c,e;
  Vec(int x,int y) {r=x;c=y;e=r+(c<<14);}
  Vec(int k) {r=k&R;c=k>>>14;e=k;}
  Vec step(int i) {return add(JO.D[i]);}
  Vec add(Vec o) {return new Vec(r+o.r,c+o.c);}
  Vec sub(Vec o) {return new Vec(r-o.r,c-o.c);}
}

class State {
  static final Vec Z = JO.Z;
  static final int R = Vec.R;
  Vec x = Z, trap;
  List<Integer> path = new ArrayList<>();
  Map<Integer,Character> G = new HashMap<>();
  Set<Integer> seen = new HashSet<>();
  Set<Integer> visited = new HashSet<>();
  State(String ln) {
    G.put(Z.e, ':');
    seen.add(Z.e);
    visited.add(Z.e);
    visit(ln);
  }

  void visit(String U) {
    visited.add(x.e);
    for (int i = 0; i < 4; ++i) {
      int p = x.step(i).e;
      if (!seen.contains(p)) {
        seen.add(p);
        G.put(p, Character.valueOf(U.charAt(i)));
      }
    }
    if (G.get(x.e) == '.')
      G.put(x.e, Character.valueOf(':'));
  }

  void printAns() {
    int c0=5000,c1=5000,r0=5000,r1=5000;
    for (int k : G.keySet()) {
      int r=k&R,c=k>>>14;
      c0=min(c0,c);
      r0=min(r0,r);
      c1=max(c1,c);
      r1=max(r1,r);
    }
    for (int k : G.keySet())
      if (G.get(k) == ':')
        G.put(k, '.');
    System.out.format("done%n%d %d%n", r1-r0+1, c1-c0+1);
    for (int r = r0; r <= r1; ++r) {
      StringBuilder sb = new StringBuilder();
      for (int c = c0; c <= c1; ++c)
        sb.append(G.getOrDefault(r+(c<<14),'#'));
      System.out.println(sb);
    }
    System.out.flush();
  }

  State go(BufferedReader in, int s) throws Exception {
    System.out.println(JO.F[s]);
    System.out.flush();
    String[] ln = in.readLine().split("\\s+");
    x = x.step(s);
    visit(ln[0]);
    if (ln[1].equals("trap")) {
      trap = x;
      return new State(ln[2]);
    }
    return null;
  }

  State search(BufferedReader in, Collection<Integer> no) throws Exception {
    while (true) {
      int s = -1;
      for (int i = 0; i < 4; ++i) {
        int p = x.step(i).e;
        if (!no.contains(p) && !visited.contains(p) && G.get(p) != '#') {
          s = i;
          break;
        }
      }
      if (s == -1) {
        if (path.isEmpty()) break;
        s = (path.get(path.size()-1)+2)%4;
        path.remove(path.size()-1);
      } else {
        path.add(s);
      }
      State succ = go(in, s);
      if (succ != null) return succ;
    }
    return null;
  }

  State walk(BufferedReader in, List<Integer> P) throws Exception {
    for (int s : P) {
      int p = x.step(s).e;
      if (G.getOrDefault(p, '#') == '#') break;
      path.add(s);
      State succ = go(in, s);
      if (succ != null) return succ;
    }
    return null;
  }

  void merge(State B, Vec O) {
    for (int k : B.G.keySet()) {
      int r=k&R,c=k>>>14;
      G.put((O.r+r)+((O.c+c)<<14), B.G.get(k));
    }
  }
}
